P036 improvement (Alignment and font can be set for each line)

uwekaditz opened the pull request:
ADD: Alignment can be set for each line (saved in DisplayLinesV1[].ModifyLayout Bits 5-3, eAlignment)
ADD: global alignment setting includes now also TEXT_ALIGN_RIGHT, paramter moved to PCONFIG_LONG(1)=P036_FLAGS_1 because it needs 2 bits
ADD: Add font Dialog_plain_18 to usable fonts
ADD: Setting and support for oledframedcmd,align,<0|1|2> subcommand, par2: (0=centre|1=left|2=right)
CHG: item DisplayLinesV1[].FontHeight renamed into DisplayLinesV1[].ModifyLayout
CHG: Preparation to modify font for each line (saved in DisplayLinesV1[].ModifyLayout Bits 2-0, eModifyFont)
CHG: Device layout uses a sub table with line number, content, modify font and alignment
CHG: Setting 'Hide indicator' is only visible for 128x64 display, its already true for all other displays

uwekaditz committed:
[P036] improvments (line alignment separatly)

ADD: global alignment setting includes now also TEXT_ALIGN_RIGHT, paramter moved to PCONFIG_LONG(1)=P036_FLAGS_1 because it needs 2 bits
CHG: item DisplayLinesV1[].FontHeight renamed into DisplayLinesV1[].ModifyLayout
CHG: Alignment can be set for each line (saved in DisplayLinesV1[].ModifyLayout Bits 5-3, eAlignment)
CHG: Preparation to modify font for each line (saved in DisplayLinesV1[].ModifyLayout Bits 2-0, eModifyFont)
CHG: Device layout uses a sub table with line number, content, modify font and alignment
CHG: Setting 'Hide indicator' is only visible for 128x64 display, its already true for all other displays
ADD: Add font Dialog_plain_18 to usable fonts
ADD: Setting and support for oledframedcmd,align,<0|1|2> subcommand, par2: (0=centre|1=left|2=right)

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -30,25 +31,43 @@ void P036_data_struct::reset() {
+# ifdef P036_FONT_CALC_LOG
+String tFontSettings::FontName() const {
+  if (fontData == ArialMT_Plain_24)
+    return "Arial_24";
```
You could have that method return a `const __FlashStringHelper *` instead of s `String`, but should at least wrap the literal strings in the regular `F("string")` (flash-string) macro, to move the string from code segment to flash.

uwekaditz replied:
Done!

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -370,20 +389,20 @@ void P036_data_struct::display_logo() {
   if (getDisplaySizeSettings(disp_resolution).PixLeft < left) { left = getDisplaySizeSettings(disp_resolution).PixLeft; }
-  top = (getDisplaySizeSettings(disp_resolution).Height - espeasy_logo_height) / 2;
+  top = (getDisplaySizeSettings(disp_resolution).Height-espeasy_logo_height)/2;
 
```
That _was_ formatted using Uncrustify, mostly for better readability 😅

uwekaditz replied:
I installed Uncrustfy in my VS Code GUI but it doesn't change the formatting.
Anything else to consider?

tonhuisman replied:
Did you also install the Uncrustify executable, and pointed the plugin at it? (It can be a bit tricky to get it going at first) Then you can select it as the default formatter (Alt-Shift-F) for VS Code.

uwekaditz replied:
I installed the Uncrustify extension and set it to my default Formatter.
But if I check the messages I got following information:
![grafik](https://user-images.githubusercontent.com/55511593/161625280-cb5a84eb-28a8-40a9-b097-ffb687ddc8eb.png)

tonhuisman replied:
Open the settings (Ctrl-,) and search for 'uncrustify', then adjust the executable path where you unpacked the [separately downloaded](https://sourceforge.net/projects/uncrustify/) uncrustify executable (yep, I'm on Windows, and temporarily using the Light+ theme 😛).

![image](https://user-images.githubusercontent.com/1922393/161628148-c87587b6-915a-4764-ae1b-0aa0d9b2758b.png)

uwekaditz replied:
Found it and it worked, see commit `Uncrustify code `

tonhuisman commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -1127,7 +1167,14 @@ void P036_data_struct::P036_DisplayPage(struct EventStruct *event)
+        switch (static_cast<eAlignment>(iAlignment)) {
+          case eAlignment::eGlobal: ScrollingPages.TxtAlignmentIn[i]=textAlignment; break;
+          case eAlignment::eLeft:   ScrollingPages.TxtAlignmentIn[i]=TEXT_ALIGN_LEFT; break;
+          case eAlignment::eCenter: ScrollingPages.TxtAlignmentIn[i]=TEXT_ALIGN_CENTER; break;
```
Duplicated code (lines 1131..1136), maybe move it to a separate method?

uwekaditz replied:
Done!

uwekaditz committed:
Changes requested by @tonhuisman

Changes requested by @tonhuisman

uwekaditz committed:
Uncrustify code

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -33,41 +33,52 @@ void P036_data_struct::reset() {
 # ifdef P036_FONT_CALC_LOG
 const __FlashStringHelper tFontSettings::FontName() const {
-  if (fontData == ArialMT_Plain_24)
+  if (fontData == ArialMT_Plain_24) {
```
Why not a switch statement here?

uwekaditz replied:
I will consider it.

uwekaditz replied:
A switch statement gives following error:
`expression must have integral or enum type`

TD-er replied:
> A switch statement gives following error: `expression must have integral or enum type`

OK, then leave it as it is... Now we also know why it isn't a switch statement ;)

uwekaditz replied:
I kept the `if endif` construction in the function FontName()
See commit `FIX: wrong alignment (only centered) for long, non-scrolling lines`

tonhuisman commented on src/_P036_FrameOLED.ino:
```diff
@@ -180,9 +189,9 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
       addLog(LOG_LEVEL_INFO, F("P036_PLUGIN_WEBFORM_LOAD ..."));
-      # endif // PLUGIN_036_DEBUG
+# endif // PLUGIN_036_DEBUG
 
```
It is partly a matter of taste, but when enabling the 'Bracket Pairs' feature of VS Code, having intermittent lines starting in column 1 is, IMHO, a bit 'disturbing', but I assume the other formatter changed that, as it once was a convention (or maybe even a requirement?) to have preprocessor directives starting in column 1.
No need to change 'm back, there are a lot of sources that have that, but Uncrustify doesn't touch them.

![image](https://user-images.githubusercontent.com/1922393/161630472-a6a0360e-79ec-4a52-8917-248f748782b4.png)

So this:
![image](https://user-images.githubusercontent.com/1922393/161630825-c8c6e1c5-a231-486f-a161-02d83c0ef2dd.png)

versus this:
![image](https://user-images.githubusercontent.com/1922393/161630878-835b1e23-01cb-47ce-80ef-d3b969b8d503.png)

TD-er replied:
> No need to change 'm back, there are a lot of sources that have that, but Uncrustify doesn't touch them.

Uncrustify will only apply formatting rules for what is defined.
So you can add it to the uncrustify config if it bothers you :)
Hint: There are online testers to test config in real time.

uwekaditz commented:
There is still a bug in this PR!
If a line is longer than the display width and line scrolling is disabled, the line is always centred independend of the alignment settings.
This bug was already in the original code of PR #3938.
I just saw it and I'm going to fix it later this week.

uwekaditz committed:
FIX: wrong alignment (only centered) for long, non-scrolling lines

FIX: wrong alignment (only centered) for long, non-scrolling lines
FIX: definition of the debug function FontName() was wrong

uwekaditz commented:
@tonhuisman and @TD-er
Can it be that we have a timing problem while writing and reading the config file including CustomTaskSettings?

- During my latest test I encountered twice the problem that all line contents were erased.

- It is also very often that after submitting the [P036] changes to the device in `PLUGIN_WEBFORM_SAVE` (including `SaveCustomTaskSettings()`)  and reloading the setting in `PLUGIN_WEBFORM_LOAD` no line content is shown anymore (`P036_data` must be a nullptr after `new (std::nothrow) P036_data_struct()` or `loadDisplayLines()` including `LoadCustomTaskSettings()` does not return to jump over the web display!)

If I'm not wrong after `PLUGIN_WEBFORM_SAVE` the `PLUGIN_INIT` (calling `loadDisplayLines()` including `LoadCustomTaskSettings()`) is called before `PLUGIN_WEBFORM_LOAD`, right?

I'm using a ESP8266 4M1M.
I suspect the longer line content strings (64 chars each) causing the upper problems.
With the shorter ones (32 chars only) I never encountered such problems.

tonhuisman commented:
> problem while writing and reading the config file

The `SaveCustomTaskSettings()` and `LoadCustomTaskSettings()` functions do take the correct measures for safe saving, by inserting enough `delay(0)` calls.
It is possible that you are reaching the max. acceptable number of writes for the flash of your unit. The counter for that can be found on the Info page, and can be reset by a power cycle. Also the age of the ESP/flash can be a factor, as the number of writes to the flash _is_ limited (exact numbers differ per manufacturer).

uwekaditz commented:
The number of flash cycles or aging of the flash is not the problem.
After reopening the device (button Close on the edit page and button Edit on the device page) the line content is shown again.
The deleted lines I could recover by manual editing. The second time I loaded a previously downloaded config file.

tonhuisman commented:
You could catch the log via USB-serial, to see if the save fails, and results in a reboot of the ESP, then that usually is caused by an out of memory exception (though you might not recognize it immediately like such). I've seen many of those in early versions om my recently merged PR on P037 MQTT Import, where I solved the memory issues by using the variants of the `*CustomTaskSettings()` functions that handle String arrays, instead of a large struct. It seems this plugin currently uses the 'large struct' variant (V1), where the V0 settings used the String array variant. You will need a data conversion on the settings if you want to switch that, by serializing all settings into a string (avoid using a comma as the separator), and parsing that on read.

uwekaditz commented:
I monitored the USB while both errors happened.
There was no reboot of the unit!

Again the question about the timing between PLUGIN_WEBFORM_SAVE, PLUGIN_INIT and PLUGIN_WEBFORM_LOAD. I could see on the display that the plug in was re-init (showing the ESP logo) between saving and reopening the web layout.

I can also check if it is a memory problem, too many `new` of the plug in structure.
But not today

tonhuisman commented:
> the plug in was re-init (showing the ESP logo) between saving and reopening the web layout.

Yes, that is by design, after `_SAVE`, the `_EXIT` event is called (always, AFAIK), and, when the plugin is enabled, again the `_INIT` event, to activate the new settings immediately. That's why it is important to properly implement the `_EXIT` event, for cleaning up any allocated memory/objects.

uwekaditz commented:
This is the free ram and stack while normal running, the edit page is not opened.
![Screenshot_20220406-224938_Firefox~01](https://user-images.githubusercontent.com/55511593/162068704-a63674a4-22a2-491f-995e-3938c522bb84.jpg)
I just powered up the unit that's why the load is so high. Now the load dropped to 10%

uwekaditz commented:
This is the free ram and stack while normal running, the edit page is not opened.
![Uploading Screenshot_20220406-224938_Firefox~01.jpg…]()

uwekaditz commented:
I'm m going to log out the ram and stack while entering plug_in_load, plug_in_save and for and after the new commands.
Let's see how it changes.

uwekaditz commented:
@tonhuisman 
I put a lot of debug messages into the code,
The lowest heap I got was
`599508 : Info   : _LOAD: After (*P036_data = new) FreeHeap:3480 FreeStack:2864`
Can it be a problem?

After just waiting with the device page opened I got even lower heap:
```
2485041 : Info   : _LOAD: Before (*P036_data = new) FreeHeap:4720 FreeStack: 2864
2485043 : Info   : _LOAD: After (*P036_data = new) FreeHeap:3480 FreeStack:2864
2485047 : Info   : _LOAD: After loadDisplayLines() FreeHeap:2888 FreeStack:2864
```
Now I got the error, again after just waiting and doing nothing, even the display was turned off.
I entered the device setting page and clicked Submit:
2859770 : Info   : _LOAD: Entering FreeHeap:4920 FreeStack:2864
2859830 : Info   : _LOAD: Before (*P036_data = new) FreeHeap:4864 FreeStack: 2864
2859837 : Info   : _LOAD: After loadDisplayLines() FreeHeap:6752 FreeStack:2864
2859909 : Info   : _LOAD: Before exit FreeHeap:6248 FreeStack:2864
2863126 : Info   : _EXIT: Before exit FreeHeap:5776 FreeStack: 2800
2863133 : Info   : _SAVE: Entering FreeHeap:9232 FreeStack:2800
2863136 : Info   : _SAVE: Before (*P036_data = new) FreeHeap:9176 FreeStack: 2800
2863136 : Info   : _SAVE: After (*P036_data = new) FreeHeap:7936 FreeStack:2800
2863197 : Info   : _SAVE: Before exit FreeHeap:9072 FreeStack:2800
2863354 : Info   : _INIT: Entering FreeHeap:8976 FreeStack:3056
2863355 : Info   : _INIT: Before (*P036_data = static_cast<P036_data_struct *>) FreeHeap:7752 FreeStack: 3056
2863423 : Info   : _INIT: After P036_data->init() FreeHeap:5560 FreeStack: 3056
2863425 : Info   : _INIT: Before exit FreeHeap:5488 FreeStack: 3056
2863737 : Info   : _LOAD: Entering FreeHeap:5416 FreeStack:2864
2865642 : Info   : _LOAD: Before (*P036_data = new) FreeHeap:3760 FreeStack: 2864

- 2865644 : Error  : P036_PLUGIN_WEBFORM_LOAD: P036_data are zero!

2865645 : Info   : _LOAD: Before exit FreeHeap:3632 FreeStack:2864

at 2865644:
- this error message is shown if `P036_data_struct *P036_data = new (std::nothrow) P036_data_struct();` returns a nullptr

This is the result:
![grafik](https://user-images.githubusercontent.com/55511593/162285091-4b87e31d-510a-49c9-bfde-9e244aacf8ea.png)

tonhuisman commented:
Your freeheap is really low, I'd normally expect it to be in the area of 10..15 kB, is this a custom build with some specific plugins or controllers included, or other memory consuming options?

uwekaditz commented:
Of course it is a custom build.
These are the includes:
 #define USES_P001   // Switch
 #define USES_P002   // ADC
 #define USES_P004   // Dallas
 #define USES_P005   // DHT
 #define USES_P023   // OLED
 #define USES_P026   // SysInfo
 #define USES_P029   // Output
 #define USES_P033   // Dummy
 #define USES_P034   // DHT12
 #define USES_P036   // FrameOLED
 #define USES_P087   // Serial Proxy
 #define USES_P092   // DLbus
 #define USES_P016   // IR
	#define P016_SEND_IR_TO_CONTROLLER false //IF true then the JSON replay solution is transmited back to the condroller.
	#define P016_P035_USE_RAW_RAW2 //Use the RAW and RAW2 encodings, disabling it saves 3.7Kb

 #define USES_C001   // Domoticz HTTP
 #define USES_C002   // Domoticz MQTT
 #define USES_C009   // FHEM HTTP

The defined devices are:
![grafik](https://user-images.githubusercontent.com/55511593/162288446-ad409065-5b47-4275-b68e-b1d5c236ccac.png)

uwekaditz commented:
Would it be more memory effective to create an extra structure for

- `tDisplayLines DisplayLinesV1[P36_Nlines];` and
- `loadDisplayLines()`

This structure can be included into `P036_data_struct`

tonhuisman commented:
Ah, the IR plugins are pretty heavy on the memory, that's why they have their own build. And also this P036 is not really light as well. You might want to try/develop/debug without the IR plugins enabled.

Best suggestion would be to try to avoid loading these (big) structures twice in memory, as that's happening for some of the plugins, AFAICS.

uwekaditz commented:
I need the IR plugin because it updates the content of the OLED display.
An extra structure as mentioned above needs only 816 bytes (68*12), the complete P036_data_struct needs 1168 bytes on the heap.

uwekaditz commented:
I'm testing a new structure, just for loading the DisplayLinesV1
```
class P036_LineContent 
{
  public:
  void loadDisplayLines(taskIndex_t taskIndex, uint8_t     LoadVersion);
  tDisplayLines DisplayLinesV1[P36_Nlines];
};

```
As expected it reduces the heap load to 860 bytes while _LOAD and _WRITE.
I will test it for a while.

Another question:
```
addTextBox( getPluginCustomArgName(varNr),
                       String(P036_data->DisplayLinesV1[varNr].Content),
                        P36_NcharsV1 - 1,
                        false,          // readonly,
                        false,          // required,
                        EMPTY_STRING,   // pattern,
                       EMPTY_STRING);  // class name

```
is not working as expected. The text box is still not maximized. but the EMPTY_STRING does work for addSelector().
![grafik](https://user-images.githubusercontent.com/55511593/162505633-7221036d-5926-4b40-ab13-2ecb67780710.png)

What is wrong?

tonhuisman commented:
> What is wrong?

Did some investigation, but it seems that it only uses the max available space if it is still less than the minimal width it would want to use. I didn't fix this either for P037 (MQTT Import) 😞 and in P104 (MAX7219 DotMatrix) the columns are much narrower, so it needs to use all available space just to have a proper input.
Without changing the CSS, probably by adding a new class `xwide` on `input` and `select` (a copy of `wide`) that has `width: 100%;`, this isn't fixable.

uwekaditz commented:
@tonhuisman `xwide` was a good idea!
I added the style `input.xwide,select.xwide{max-width:500px;width:95%}` as hex code to
`static const char DATA_ESPEASY_DEFAULT_MIN_CSS[]`
in WebStaticData.h and it looks now good:
![grafik](https://user-images.githubusercontent.com/55511593/162580112-fd239d00-b0fb-4842-abd2-8610e4e8517f.png)

tonhuisman commented:
> as hex code to
> `static const char DATA_ESPEASY_DEFAULT_MIN_CSS[]`

Then please also add that to `static\espeasy_default.css`, as that's the base file used for generating that hex structure (by running `embed_files.sh`)

uwekaditz commented:
I did it already!
I tested also the width of 100% but it is to narrow to the next field:
![grafik](https://user-images.githubusercontent.com/55511593/162580552-ae506c04-1071-4830-8e8d-8e77fea580f0.png)
I'm not so familiar with html programming, so my question:
Can a margin and 100% be used for this style?

uwekaditz commented:
I found the solution for a wider gap with style xwide and 100%:
The data cell holding the textbox must be created with
`<td style="padding-left:8px;padding-right:8px;">`
Should it be includes as a special `html_TD()` function, e.g. `html_TD(const String& classname)`?

uwekaditz committed:
FIX: wrong pixel calculation for alignment for non-scrolling lines on 64x48 display

FIX: wrong pixel calculation for alignment for non-scrolling lines on 64x48 display (the first shown left pixel on 64x48 displays is 32!)
CHG: new class P036_LineContent(), just for loading the DisplayLinesV1

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
[P036] NEW: individual font settings for each line

NEW: individual font settings for each line
FIX: while setting new 'Lines per Frame' by 'oledframedcmd' the max page count was not updated
FIX: 'Tweaked to match the 13 pix font to fit for 4 lines display' did only work for 128x64 displays
CHG: 'Turn on/off the Indicator if the number of frames changes' removed because it is not needed
MSG: Macros for P036_DisplayIsOn and P036_SetDisplayOn
MSG: Preperation for css style xwide (define P036_USE_XWIDE)

uwekaditz commented:
Just a note to my last commit:
The line 448 `html_TD(F("padding-right: 8px"));` in _P036_FrameOLED.ino
requires a new prototype in WebServer/HTML_wrappers.cpp
```
void html_TD(const String& style) {
  addHtml(F("<TD style=\""));
  addHtml(style);
  addHtml(F(";\">"));
}
```
that I havened committed!

tonhuisman commented:
> that I havened committed!

Now it breaks the build, so you should just commit those files too, git will usually merge it nicely.

uwekaditz commented:
It does not brake this build because it is wrapped with a `#ifdef...#endif` directive!

Should I add the changes for the `xwide `style to this PR or make a new one?
By the way two questions:

- Why is `DATA_ESPEASY_DEFAULT_MIN_CSS` coded with hex numbers and not in ascii as other definitions?
- Who creates the hex code for `DATA_ESPEASY_DEFAULT_MIN_CSS` from the \static\espeasy_default.css

tonhuisman commented:
> It does not brake this build because it is wrapped with a `#ifdef...#endif` directive!

Sorry, I hadn't looked closely at the breaking build yet 😞 that's a bin size issue, as so often

>  Who creates the hex code for `DATA_ESPEASY_DEFAULT_MIN_CSS` from the \static\espeasy_default.css

There is a script `embed_files.sh` that minifies all js and css code into `static\data_h_temp` and that is manually updated in ` src\src\Static\WebStaticData.h`. Not sure, but I'd expect anyone updating the css to include that in the PR, but I assume it can also be done automagically during the release build. That part is a bit unclear to me, atm.

uwekaditz commented:
- Should I try to reduce the code, just 50 bytes brake the the build for `esp8266, normal_ESP8266_1M`?

- What is with the `xwide `style and the `html_TD(const String& style)`?
Should I add the changes to this PR or make a new one?

uwekaditz committed:
code reduced by 530 bytes to fit the build size for 'esp8266, normal_ESP8266_1M'

equal code put into functions

uwekaditz commented:
Attached are 3 pictures for a 128x64 display with 3 lines per frame showing the latest changes:
![128x64 page3](https://user-images.githubusercontent.com/55511593/163843751-a78274ab-d142-44a6-a4c0-88abc7c24550.jpg) ![128x64 page2](https://user-images.githubusercontent.com/55511593/163843757-4e6888c2-c792-4503-bc96-5b82aa825442.jpg)
![128x64 page1](https://user-images.githubusercontent.com/55511593/163843760-6b5aab95-e48d-4727-b2d6-2354253b819a.jpg)

uwekaditz commented:
@tonhuisman I have another idea for an alignment `left|right` to display two parameters in the same line, but one left the other one right aligned.
What I need is a suggestion for a token that defines the split, e.g. `<split>` or similar.

tonhuisman commented on src/_P036_FrameOLED.ino:
```diff
@@ -1120,6 +1123,27 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+            String precharge = parseString(string, 5);
+            String comdetect = parseString(string, 5);
+            success = true;
+            P036_data->display->setContrast(contrast.toInt(),precharge.toInt(),comdetect.toInt());
```
Why would you create `String` objects using 3 rather costly `parseString()` calls, and parse `int`s from these if the values are already available in `event->Par2..4`? It will be much cheaper, and memory-friendlier, to use `static_cast<some_type>()` these if they don't match according to the compiler.

uwekaditz replied:
I used the code for the other commands as master copy.
I was not sure if a static cast to the event parameters will work if the user enters non-number characters.

uwekaditz replied:
Changed!
See my commit [45fe0fe](https://github.com/letscontrolit/ESPEasy/pull/3997/commits/45fe0fe78dbebdb5b1394977e4dc2e9f2cc745d4)

tonhuisman commented:
> What I need is a suggestion for a token that defines the split, e.g. `<split>` or similar.

I usually try to find a character/sequence that is not (often) used, and preferably, short, but for a display, that's quite difficult. I'd probably end up with smth. like `<|>` or `<$>`, or a character that can't be displayed by the font, like `█` (alt+219), but I'm not sure how that will be stored in the settings, nor can it be easily entered from a non-Windows device.

uwekaditz commented:
I like the `<|>`.

uwekaditz committed:
ADD: Setting and support for user defined contrast: oledframedcmd,display,user,contrast,precharge,comdetect

ADD: oledframedcmd,display,user,contrast,precharge,comdetect (contrast, precharge and comdetect are integers)
CHG: setting for low contrast modified, low was invisible!

uwekaditz committed:
Copy&paste error

uwekaditz committed:
CHG: Setting for user defined contrast simplified

CHG: Setting for user defined contrast simplified (parse int from event->Par3..5)

uwekaditz commented:
Use of the split token `<|>`:
The line settings
line 7: `left<|>right, None, center`
line 8: `2, Use biggest, center`
are displayed on a 128x64 display as shown:
![grafik](https://user-images.githubusercontent.com/55511593/164091961-c155390e-7c07-45a2-bd10-41874b876a27.png)

uwekaditz commented:
Another example looks like
![grafik](https://user-images.githubusercontent.com/55511593/164093980-d07a3060-cd5b-47a1-8c70-dc247cbb707f.png)

as requested in #1711 (Improve OLED Framed formatting configuration, including text alignment)

uwekaditz committed:
NEW: use the split token <|> to split lines into left and right part

- the split token is replaced by a number of space chars to fit the display width
- if the modified line is longer than the display width (even with only one space as replacement), the modified line will be scrolled, if line scrolling is enabled
- no need to use a special alignment for displaying the line left and right aligned

uwekaditz commented:
I created a new PR #4010 for the css style xwide.

uwekaditz committed:
FIX: last frame was not shown

frameCounter check was wrong

uwekaditz committed:
Documentation updated

TD-er committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

TD-er committed:
Merge branch 'mega' into P036_improvement

TD-er commented:
One small bug...
I have it setup with 2 lines and 2 lines per page. (no header)
However, it switches between showing those 2 lines and a blank screen. (128x32 pixel screen)

![image](https://user-images.githubusercontent.com/3751318/164803256-8c45c4b2-b99d-41f6-829e-7e78462301d7.png)

TD-er commented:
Just some heads up.
I got an idea to reduce the memory usage when loading/saving a task with this plugin.
Will push it later to this PR.

TD-er committed:
[Framed OLED] Load/save settings in smaller memory chunks

This does make it more likely memory allocation will succeed instead of large chunks at once.

uwekaditz committed:
FIX: empty page was shown if just one frame has text

The while loop in P036_DisplayPage() was left too early if just one frame has text to display

uwekaditz committed:
FIX: empty page was shown if just one frame has text

The while loop in P036_DisplayPage() was left too early if just one frame has text to display

TD-er committed:
[Framed OLED] Reduce memory usage, especially during load/save

No need to keep the raw data structure in memory, as this structure reserves the maximum allowed string length, where often not all strings are used and for sure not all at their max. length.

When loading/saving, only a small part of this raw structure needs to be active in memory.

TD-er committed:
Merge remote-tracking branch 'uwekaditz/P036_improvement' into P036_improvement

TD-er commented:
I added the intended optimizations to this pull request.
@uwekaditz Please do a git fetch and git pull before you continue working on it.

I will test your fix on showing the empty page.

TD-er commented:
There is still something strange going on with memory usage.
But it is not related to this plugin, it seems to be related to how the webserver seems to work internally.

When saving a page, you send quite a large piece of data to the ESP via HTTP POST.
This seems to be kept in memory until another call to the web server is made (or nr of calls, or maybe a new POST, no idea yet). It will also be freed on its own after some time.

Anyway, the amount of data sent when saving something in this plugin, is quite large. This makes this plugin a bit more sensitive to this effect.
So I have to look into clearing these buffers in the webserver to free this memory before trying to call the init of this plugin as this init also needs quite a lot of memory. (well now since my last commit, it is needing less memory)

TD-er committed:
[OLED Framed] Prevent copy of strings when calling some write functions

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -783,40 +1050,99 @@ uint8_t P036_data_struct::display_scroll(ePageScrollSpeed lscrollspeed, int lTas
-        log += F(" PixLength: "); log += display->getStringWidth(ScrollingPages.LineIn[j]);
+        log += F(" -> Changed to: "); log += ScrollingPages.In[j].SPLcontent;
+        log += F(" Length: "); log += ScrollingPages.In[j].SPLcontent.length();
+        display->setFont(FontSizes[LineSettings[ScrollingPages.In[j].SPLidx].fontIdx].fontData);
```
I think this is an error.
You perform actions which depend on the set log level.

uwekaditz replied:
It's not an error.
display->setFont() must be used to calculate the display->getStringWidth() according to the individual font.
The StringWidth should just be logged, nothing else.
Later the setting of display->setFont() is overwritten by the correct values for this line.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -829,12 +1155,13 @@ uint8_t P036_data_struct::display_scroll(ePageScrollSpeed lscrollspeed, int lTas
-        log += F(" PixLength: "); log += display->getStringWidth(ScrollingPages.LineOut[j]);
+        log += F(" -> Changed to: "); log += ScrollingPages.Out[j].SPLcontent;
+        log += F(" Length: "); log += ScrollingPages.Out[j].SPLcontent.length();
+        display->setFont(FontSizes[LineSettings[ScrollingPages.Out[j].SPLidx].fontIdx].fontData);
```
Idem, performing actions depending on the set log level.

uwekaditz replied:
It's not an error.
display->setFont() must be used to calculate the display->getStringWidth() according to the individual font.
The StringWidth should just be logged, nothing else.
Later the setting of display->setFont() is overwritten by the correct values for this line.

TD-er commented:
Can you also make the inclusion of Dialog_plain_18 optional, so it will not be included when LIMIT_BUILD_SIZE is defined?
Not sure if this will break other code, so that's why I have not yet done it.
But right now the normal 1M build fails and this font is 7.2 k in size, so it may make the difference here.

uwekaditz committed:
CHG: font Dialog_plain_18 is optional (not using LIMIT_BUILD_SIZE)

uwekaditz committed:
CHG: font Dialog_plain_18 is optional (not used with LIMIT_BUILD_SIZE)

Forgot to exclude
# include <Dialog_Plain_18_font.h>
for LIMIT_BUILD_SIZE

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
CHG: Logging in [P092] reduced for LIMIT_BUILD_SIZE

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

tonhuisman commented:
Other measures are needed to remove the ~3 kB that the build needs to be slimmed down. Rebuilding like this will keep failing...
Maybe you can disable one of the fonts if `LIMIT_BUILD_SIZE` is #defined ?

```
RAM:   [======    ]  57.7% (used 47288 bytes from 81920 bytes)
Flash: [==========]  100.3% (used 1047556 bytes from 1044464 bytes)
Error: The program size (1047556 bytes) is greater than maximum allowed (1044464 bytes)
```

uwekaditz commented:
The failing build `test_D_alt_wifi_ESP8266_4M1M_VCC` is not using `LIMIT_BUILD_SIZE`!
I have already reduced the logging and removed one font for `LIMIT_BUILD_SIZE`.

tonhuisman commented:
I think it does, from `define_plugin_sets.h`:
```
// TESTING #####################################
#ifdef PLUGIN_SET_TESTING
  #if !defined(PLUGIN_SET_MAX) && !defined(ESP32)
    #ifndef LIMIT_BUILD_SIZE
      #define LIMIT_BUILD_SIZE
    #endif
```

The alternate WiFi stuff probably includes some extra debugging, causing this undesired increase of build size 😢

uwekaditz commented:
I'll check it with my VS environment.

uwekaditz committed:
CHG: font Dialog_plain_12 is optional (not used with LIMIT_BUILD_SIZE)

uwekaditz committed:
P36_MaxFontCount reduced for LIMIT_BUILD_SIZE

TD-er commented:
@tonhuisman Have you tested this too?

tonhuisman commented:
> @tonhuisman Have you tested this too?

Not yet, but I will start testing soon

uwekaditz commented:
I tested the new functions with the 3 displays supported by P036, 128x64,  128x32 and 64x48 using the SSD1306 controller.
I do not have a SH1106 controller, so I couldn't test it.

tonhuisman commented:
I've been testing with dual SSD1306 controllers (on a single ESP8266), and it seems to work mostly fine, though I have a few suggestions for UI improvement and a possible alignment issue (below):
1) UI suggestions:
- Move the Lines sub-header above the Align content (global) combo
- Add a RowLabel before the table, and remove the word Line from the line # column
  - This prevents the table to mess up the page layout, and again shows the Interval setting as we are used to

Screenshot:
![Screenshot - 08-05-2022 , 14_11_08](https://user-images.githubusercontent.com/1922393/167295931-c5c2e201-8532-4111-a711-ee7e231211c4.png)

The column/rowlabel fix in code:
```cpp
          addRowLabel(F("Line"));
          
          html_table(EMPTY_STRING); // Sub-table
          html_table_header(F("&nbsp;#&nbsp;"));
```

2) Possible layout issue:
As you can see in the screenshot my content for line 1 has a couple of spaces, and the digit 1, to test the left-align feature to keep these spaces, when I added left-align before. With the current code, this 1 is aligned fully to the left of the display, when left aligned (`trim()`?), and thus prevents to correct content alignment by adding spaces. Can that previous feature be restored/fixed?
I do understand this is a bit harder to fix for right-aligned content, but you could add a setting for a replacement character (underscore?) to be transformed into space, just before displaying the content.

uwekaditz committed:
FIX: left alignment again with leading spaces

Comment: the leading spaces were trimmed because P36_parseTemplate() did only check the global alignment setting but not the indivuidual setting for this line
CHG: UI suggestions from tonhuisman

uwekaditz commented:
This is the result of the UI changes:
![changed](https://user-images.githubusercontent.com/55511593/167305590-13b75792-bfaf-4da1-914b-5b86e187a2b2.JPG)

uwekaditz committed:
Merge branch 'mega' into P036_improvement

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -969,46 +1192,58 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
-          P036_data->DisplayLinesV1[LineNo - 1].reserved                  = (event->Par3 & 0xFF); // not implemented yet
+          String* currentLine = &P036_data->LineContent->DisplayLinesV1[LineNo - 1].Content;
+          *currentLine = parseStringKeepCase(string, 3);
+          *currentLine = P036_data->P36_parseTemplate(*currentLine, LineNo - 1);
```
I don't get it why the minimal line length should be based on the line nr.

uwekaditz replied:
The function `P36_parseTemplate()` has the `minimal_lineSize` of 20 now hard coded inside.
The `lineIdx` (LineNo-1) is necessary to get the alignment for this line to handle the leading spaces for left aligned lines as requested by tonihuisman.
I changed the prototype for this function!

uwekaditz commented:
@tonhuisman Would it be useful to have the same handling as you requested for leading spaces (left aligned) for trailing spaces on right aligned lines?

tonhuisman commented:
> same handling as you requested for leading spaces (left aligned) for trailing spaces on right aligned lines?

That sounds useful, indeed. I've been thinking about that already, but didn't express it yet.

uwekaditz commented:
Ok, than I'll added it with my next commit.

uwekaditz committed:
NEW: right alignment with trailing spaces

P36_parseTemplate():
immediately returns EMPTY_STRING if original string is empty (nothing to parse)
TEXT_ALIGN_LEFT: add leading spaces from original string to result
TEXT_ALIGN_RIGHT: add trailing spaces from original string to result

tonhuisman commented:
That is all working as intended, AFAICS. I can even combine left align or right align with a split `<|>` and it aligns as is expected (though that's an unusual combination), keeping any spaces left or right.
Looking great 🥇

uwekaditz commented:
@tonhuisman Thanks for testing!

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz commented:
With the two latest merge of 'mega' into this PR the build (esp8266, test_D_alt_wifi_ESP8266_4M1M_VCC) failed because of this:
`Error: The program size (1045128 bytes) is greater than maximum allowed (1044464 bytes)`

I was so happy that I could trim my code before this two merges that it fitted all builds! :disappointed_relieved:

tonhuisman commented:
Size on that specific build has been a challenge for a long time...

uwekaditz committed:
Merge branch 'mega' into P036_improvement

tonhuisman commented:
@uwekaditz Can you please make a small mod to the documentation of this plugin, as some user(s) [feel a bit mislead or confused](https://www.letscontrolit.com/forum/viewtopic.php?t=9161) by this sentence:
`The plugin allows for a header and footer line, which may show some user selectable information.`

I would like to suggest to change it to something like this:
`The plugin allows for a header line, which may show some user selectable information. And an optional footer line that can show a page indicator, when using a 128x64 display.`

uwekaditz committed:
Update P036.rst

Documentation for header and footer changed according to suggestion from  tonhuisman

TD-er committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

TD-er committed:
Merge branch 'mega' into P036_improvement

TD-er committed:
Merge branch 'mega' into P036_improvement

tonhuisman commented on src/_P036_FrameOLED.ino:
```diff
@@ -218,8 +278,12 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
-
+# ifdef P036_ENABLE_LEFT_ALIGN
+      {
+        addFormCheckBox(F("Reduce line number to fit font"), F("p036_ReduceLineNo"), bitRead(P036_FLAGS_1, P036_FLAG_REDUCE_LINE_NO));
+        addFormNote(F("When checked, 'Lines per Frame' will be automatically reduced to fit the individual line settings."));
```
2 remarks here:
a) 'Reduce line number...' is wonky English, could be 'Reduce number of lines...' or 'Reduce nr. of lines...'

b) To reduce build size, the `addFormNote()` calls could be wrapped in `#ifndef LIMIT_BUILD_SIZE`, as the information shown is helpful but should *also* be reflected in the documentation. (This applies multiple times)

uwekaditz replied:
Done!

uwekaditz committed:
Changes proposed and requested by tonhuisman

CHG: using css style xwide as default (define P036_USE_XWIDE deleted)
CHG: correct English text (requested by tonhuisman)
CHG: addFormNote() not used with LIMIT_BUILD_SIZE (proposed by tonhuisman)
CHG: Documentation updates

uwekaditz committed:
History updated

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Added new sub-table from PR #4239

uwekaditz committed:
CHG: use P036_LIMIT_BUILD_SIZE if PLUGIN_BUILD_IR is defined

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz commented:
@tonhuisman Finally, I managed to keep this PR with the new features within the limited build size. 😄

tonhuisman commented:
I was going to suggest that last change, but I expected you to figure that one out yourself 👍

uwekaditz committed:
Merge branch 'mega' into P036_improvement

TD-er commented:
> @tonhuisman Finally, I managed to keep this PR with the new features within the limited build size. 😄

Now you know how much work and how frustrating it really is to get it all to fit in a build. :)
Join the club!

TD-er commented:
By the way, I was just browsing through your code, to see whether I should merge this before or after the next upcoming build...
Then I came across this:
```c++
#ifdef INPUT_PULLDOWN
```
Where should this be defined? Is it part of the ESP32 code, as ESP8266 doesn't have an option to choose between pull-up or pull-down. A pin either has a pull-up or a pull-down resistor (or none).
Also I'm not sure if such specific code should be part of a specific plugin, or perhaps more central as an option for every plugin which has the option to enable a pull-up.

Another option is to set this on the Hardware tab as you have all options there which do know about the available options per pin as also on ESP32 not all pins have the same nr of options.

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz commented:
@tonhuisman Thanks for trusting in my skills 😉

uwekaditz commented:
@TD-er  `#ifdef INPUT_PULLDOWN` was obsolete code!
I'll remove it with my next commit.

uwekaditz committed:
Obsolete code with #ifdef INPUT_PULLDOWN removed

Changes in P036 and P092

uwekaditz committed:
Merge branch 'P036_improvement' of https://github.com/uwekaditz/ESPEasy into P036_improvement

uwekaditz commented:
@TD-er 

> Also I'm not sure if such specific code should be part of a specific plugin, or perhaps more central as an option for every plugin which has the option to enable a pull-up.

It would be a good idea, if there is a general function that a plugin can use its own pin mode settings and may be overwrite the genereal settings from the hardware page.
Another possibility would be that the plugin is not able to set its own pin mode but there is an additional note that reminds to set the pin mode on the hardware page.

TD-er commented:
> It would be a good idea, if there is a general function that a plugin can use its own pin mode settings and may be overwrite the genereal settings from the hardware page.

Sure, but that should be done for all plugins, not implemented in the plugin code as that will lead to code duplication.
And code duplication is bad for so many reasons, with the most important ones being:
- When fixing/changing something it must be done at all places or else there will be remaining bugs or different behavior at different places.
- Bin size will be larger than needed
- It gives me some allergic reaction, with lots of itching ;)

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
BugFix for addSelector()

In the latest merge of branch 'mega' into P036_improvement the parameter class name in function AddSelector() has changed from EMPTY_STRING to  F("")
_P036_FrameOLED.ino has not been updated while this change was merged to mega

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Conflict resolved

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Conflicts resolved

uwekaditz commented:
@tonhuisman  I merged the latest mega to this PR.
While building `Build / build (esp8266, display_ESP8266_4M1M) (pull_request) ` the check failed with following error:

> src/src/PluginStructs_tmp/__tmpfile.cpp:17298:36: warning: deleting object of polymorphic class type 'Adafruit_PCD8544' which has non-virtual destructor might cause undefined behaviour [-Wdelete-non-virtual-dtor]
   if (nullptr != pcd8544) { delete pcd8544; }
                                    ^
In file included from src/src/PluginStructs_tmp/../PluginStructs/P141_data_struct.h:8:0,
                 from src/src/PluginStructs_tmp/__tmpfile.cpp:17134:
lib/esp8266-oled-ssd1306/OLEDDisplay.h: At global scope:
.pio/libdeps/display_ESP8266_4M1M/Adafruit PCD8544 Nokia 5110 LCD library/Adafruit_PCD8544.h:30:15: error: expected identifier before numeric constant
 #define BLACK 1 ///< Black pixel
               ^
lib/esp8266-oled-ssd1306/OLEDDisplay.h:102:3: note: in expansion of macro 'BLACK'
   BLACK = 0,
   ^
.pio/libdeps/display_ESP8266_4M1M/Adafruit PCD8544 Nokia 5110 LCD library/Adafruit_PCD8544.h:30:15: error: expected '}' before numeric constant
 #define BLACK 1 ///< Black pixel
               ^
lib/esp8266-oled-ssd1306/OLEDDisplay.h:102:3: note: in expansion of macro 'BLACK'
   BLACK = 0,
   ^
.pio/libdeps/display_ESP8266_4M1M/Adafruit PCD8544 Nokia 5110 LCD library/Adafruit_PCD8544.h:30:15: error: expected unqualified-id before numeric constant
 #define BLACK 1 ///< Black pixel
               ^
lib/esp8266-oled-ssd1306/OLEDDisplay.h:102:3: note: in expansion of macro 'BLACK'
   BLACK = 0,
   ^
In file included from lib/esp8266-oled-ssd1306/SSD1306Wire.h:31:0,
                 from lib/esp8266-oled-ssd1306/SSD1306.h:30,
                 from src/src/PluginStructs_tmp/../PluginStructs/P036_data_struct.h:7,
                 from src/src/PluginStructs_tmp/__tmpfile.cpp:20562:
lib/esp8266-oled-ssd1306/OLEDDisplay.h:105:1: error: expected declaration before '}' token
 };
 ^
*** [.pio/build/display_ESP8266_4M1M/src/src/PluginStructs_tmp/__tmpfile.cpp.o] Error 1

What is wrong?

tonhuisman commented:
> What is wrong?

It appears that there is a conflicting definition for `BLACK` used in `OLEDDisplay.h` and `Adafruit_PCD8544.h`.
The weird part here is that this has _never_ tripped in the P141 (PCD8544 display) builds, and of course my local build just runs fine, but the Windows compiler is somewhat more relaxed than the Linux compiler, that's a know 'issue'.
Don't yet know how to solve this, though, other than renaming either of the definitions.

I have an open [PR](https://github.com/adafruit/Adafruit-PCD8544-Nokia-5110-LCD-library/pull/65) at the Adafruit_PCD8544 library for fixing that other issue of the destructor not being virtual, created ca. 2 months ago, that didn't get _any_ response yet 😞 so I can either pull that library local to ESPEasy and fix these issues, or wait 'forever' for these changes to be merged, so pulling that library local seems the easiest next step.

tonhuisman commented:
I've created PR #4297 to resolve these issues.

TD-er committed:
Merge branch 'mega' into P036_improvement

uwekaditz commented:
@TD-er : By the way, what is the forecast for this PR? Will it be soon merged into mega?
While merging mega in this PR I got conflicts, because the official sources for P036 and P092 on mega are out-of-date.

TD-er commented:
> @TD-er : By the way, what is the forecast for this PR? Will it be soon merged into mega? While merging mega in this PR I got conflicts, because the official sources for P036 and P092 on mega are out-of-date.

As you can see, I just merged the 'mega' branch into this PR and it built successfully.
So what kind of merge issues did you experience?

uwekaditz commented:
I mean the merges in the last 23 days.
I always added 'Conflict resolved' or ' Bugfix' while merging.

tonhuisman commented:
@uwekaditz When fixing the merge issues (sorry for that), can you also change in `P036_data_struct.h` the `uint8_t displayTimer    = 0;` from `uint8_t` to `uint16_t` so it will hold display time-out values over 255? (65535 seconds should be plenty)

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
CHG: Display timeout is now a uint16_t value (max 65535s for display off)

uwekaditz commented:
Display timeout is now a uint16_t value (max 65535s for display off, see #4295

uwekaditz committed:
Merge branch 'mega' into P036_improvement

uwekaditz committed:
Merge branch 'mega' into P036_improvement
